rpc: TestRemoteOffsetUnhealthy failed

rpc.TestRemoteOffsetUnhealthy [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16106696?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/16106696?buildTab=artifacts#/) on release-24.1 @ [c2e1ae99aca6c6b59b2e078e423a1148ad4c8d45](https://github.com/cockroachdb/cockroach/commits/c2e1ae99aca6c6b59b2e078e423a1148ad4c8d45):


```
=== RUN   TestRemoteOffsetUnhealthy
I240717 06:41:15.621861 429 2@rpc/peer.go:524  [rnode=2,raddr=127.0.0.1:36733,class=default,rpc] 19  connection is now healthy
I240717 06:41:15.627213 472 2@rpc/peer.go:524  [rnode=3,raddr=127.0.0.1:46049,class=default,rpc] 20  connection is now healthy
I240717 06:41:15.635963 460 2@rpc/peer.go:524  [rnode=4,raddr=127.0.0.1:33439,class=default,rpc] 21  connection is now healthy
I240717 06:41:15.641771 477 2@rpc/peer.go:524  [rnode=1,raddr=127.0.0.1:38327,class=default,rpc] 22  connection is now healthy
I240717 06:41:15.649856 508 2@rpc/peer.go:524  [rnode=3,raddr=127.0.0.1:46049,class=default,rpc] 23  connection is now healthy
I240717 06:41:15.655704 487 2@rpc/peer.go:524  [rnode=4,raddr=127.0.0.1:33439,class=default,rpc] 24  connection is now healthy
I240717 06:41:15.661218 563 2@rpc/peer.go:524  [rnode=1,raddr=127.0.0.1:38327,class=default,rpc] 25  connection is now healthy
I240717 06:41:15.666486 575 2@rpc/peer.go:524  [rnode=2,raddr=127.0.0.1:36733,class=default,rpc] 26  connection is now healthy
I240717 06:41:15.672296 590 2@rpc/peer.go:524  [rnode=4,raddr=127.0.0.1:33439,class=default,rpc] 27  connection is now healthy
I240717 06:41:15.683630 395 2@rpc/peer.go:524  [rnode=1,raddr=127.0.0.1:38327,class=default,rpc] 28  connection is now healthy
I240717 06:41:15.689374 280 2@rpc/peer.go:524  [rnode=2,raddr=127.0.0.1:36733,class=default,rpc] 29  connection is now healthy
I240717 06:41:15.694821 622 2@rpc/peer.go:524  [rnode=3,raddr=127.0.0.1:46049,class=default,rpc] 30  connection is now healthy
E240717 06:41:15.992796 622 2@rpc/peer.go:577  [rnode=3,raddr=127.0.0.1:46049,class=default,rpc] 31  disconnected (was healthy for 0s): grpc: context deadline exceeded [code 4/DeadlineExceeded]
E240717 06:41:15.993001 508 2@rpc/peer.go:577  [rnode=3,raddr=127.0.0.1:46049,class=default,rpc] 32  disconnected (was healthy for 0s): operation "conn heartbeat" timed out after 167ms (given timeout 100ms): grpc: context deadline exceeded [code 4/DeadlineExceeded]
    context_test.go:1203: max offset: 100ms - node 0 with acceptable clock offset of 0s did not return an error, as expected
    context_test.go:1201: max offset: 100ms - node 1 with acceptable clock offset of 0s returned unexpected error: clock synchronization error: this node is more than 100ms away from at least half of the known nodes (1 of 2 are within the offset)
    context_test.go:1203: max offset: 100ms - node 2 with acceptable clock offset of 0s did not return an error, as expected
    context_test.go:1195: max offset: 100ms - node 3 with excessive clock offset of 100.000001ms returned expected error: clock synchronization error: this node is more than 100ms away from at least half of the known nodes (0 of 2 are within the offset)
--- FAIL: TestRemoteOffsetUnhealthy (0.42s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRemoteOffsetUnhealthy.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: CRDB-40377

--- a/pkg/rpc/context_test.go
+++ b/pkg/rpc/context_test.go
@@ -1104,8 +1104,12 @@ func TestRemoteOffsetUnhealthy(t *testing.T) {
 		clock := timeutil.NewManualTime(timeutil.Unix(0, start.Add(nodeCtxs[i].offset).UnixNano()))
 		nodeCtxs[i].errChan = make(chan error, 1)
 		nodeCtxs[i].ctx = newTestContext(clusterID, clock, maxOffset, stopper)
-		nodeCtxs[i].ctx.RPCHeartbeatInterval = maxOffset
-		nodeCtxs[i].ctx.RPCHeartbeatTimeout = maxOffset
+		// Make the test faster.
+		nodeCtxs[i].ctx.RPCHeartbeatInterval = 10 * time.Millisecond
+		// Disable RPC heartbeat timeouts to avoid flakiness in the test. If a
+		// heartbeat were to time out, its RPC connection would be closed and its
+		// clock offset information would be lost.
+		nodeCtxs[i].ctx.RPCHeartbeatTimeout = 0
 		nodeCtxs[i].ctx.NodeID.Set(context.Background(), roachpb.NodeID(i+1))
 
 		s := newTestServer(t, nodeCtxs[i].ctx)

